File: Proyecto-Factura-Contabilidad/src/app/servicios/tipo-pago.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment.development';
import { ApiListResponse, ApiResponse } from '../core/models/api-response';

export interface TipoPagoListadoDto {
  tpaId: number;
  tpaNombre: string;
  createdAt: string;
  updatedAt: string | null;
}

export interface TipoPagoCrearDto {
  tpaNombre: string;
}

export interface TipoPagoEditarDto extends TipoPagoCrearDto {}

export interface TipoPagoFiltroDto {
  nombre?: string;
  pagina?: number;
  registrosPorPagina?: number;
}

@Injectable({ providedIn: 'root' })
export class TipoPagoService {
  private http = inject(HttpClient);
  private readonly API = environment.transaccionesApiUrl;

  listar(filtro?: TipoPagoFiltroDto): Observable<ApiListResponse<TipoPagoListadoDto>> {
    let params = new HttpParams();

    if (filtro?.nombre) params = params.set('nombre', filtro.nombre);
    if (filtro?.pagina != null) params = params.set('pagina', filtro.pagina);
    if (filtro?.registrosPorPagina != null) {
      params = params.set('registrosPorPagina', filtro.registrosPorPagina);
    }

    return this.http.get<ApiListResponse<TipoPagoListadoDto>>(`${this.API}/api/TipoPago`, { params });
  }

  obtenerPorId(id: number): Observable<ApiResponse<TipoPagoListadoDto>> {
    return this.http.get<ApiResponse<TipoPagoListadoDto>>(`${this.API}/api/TipoPago/${id}`);
  }

  crear(dto: TipoPagoCrearDto): Observable<ApiResponse<TipoPagoListadoDto>> {
    return this.http.post<ApiResponse<TipoPagoListadoDto>>(`${this.API}/api/TipoPago`, dto);
  }

  editar(id: number, dto: TipoPagoEditarDto): Observable<ApiResponse<any>> {
    return this.http.put<ApiResponse<any>>(`${this.API}/api/TipoPago/${id}`, dto);
  }

  eliminar(id: number): Observable<ApiResponse<any>> {
    return this.http.delete<ApiResponse<any>>(`${this.API}/api/TipoPago/${id}`);
  }
}

File: Microservicio.Transacciones.Api/Services/TipoPagoService.cs
using Microsoft.EntityFrameworkCore;
using Microservicio.Transacciones.Api.Data;
using Microservicio.Transacciones.Api.DTOs.TipoPago;
using Microservicio.Transacciones.Api.Interfaces;
using Microservicio.Transacciones.Api.Models;

namespace Microservicio.Transacciones.Api.Services;

public class TipoPagoService : ITipoPagoService
{
    private readonly TransaccionesDbContext _context;

    public TipoPagoService(TransaccionesDbContext context)
    {
        _context = context;
    }

    public async Task<(IEnumerable<TipoPagoListadoDto> Items, int TotalRegistros)> ListarAsync(TipoPagoFiltroDto filtro)
    {
        filtro.Pagina = filtro.Pagina <= 0 ? 1 : filtro.Pagina;
        filtro.RegistrosPorPagina = filtro.RegistrosPorPagina <= 0 ? 10 : filtro.RegistrosPorPagina;

        var query = _context.TipoPagos
            .AsNoTracking()
            .Where(t => t.DeletedAt == null)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(filtro.Nombre))
        {
            var nombre = filtro.Nombre.Trim();
            query = query.Where(t => t.TpaNombre.Contains(nombre));
        }

        var totalRegistros = await query.CountAsync();

        var items = await query
            .OrderByDescending(t => t.TpaId)
            .Skip((filtro.Pagina - 1) * filtro.RegistrosPorPagina)
            .Take(filtro.RegistrosPorPagina)
            .Select(t => new TipoPagoListadoDto
            {
                TpaId = t.TpaId,
                TpaNombre = t.TpaNombre,
                CreatedAt = t.CreatedAt,
                UpdatedAt = t.UpdatedAt
            })
            .ToListAsync();

        return (items, totalRegistros);
    }

    public async Task<TipoPagoListadoDto?> ObtenerPorIdAsync(int id)
    {
        var tipoPago = await _context.TipoPagos
            .AsNoTracking()
            .FirstOrDefaultAsync(t => t.TpaId == id && t.DeletedAt == null);

        if (tipoPago == null)
            return null;

        return MapearTipoPago(tipoPago);
    }

    public async Task<TipoPagoListadoDto> CrearAsync(TipoPagoCrearDto dto, int? createdBy = null)
    {
        var nombre = dto.TpaNombre.Trim();

        var existe = await _context.TipoPagos
            .AnyAsync(t => t.TpaNombre == nombre && t.DeletedAt == null);

        if (existe)
            throw new Exception("Ya existe un tipo de pago con ese nombre.");

        var tipoPago = new TipoPago
        {
            TpaNombre = nombre,
            CreatedAt = DateTime.Now,
            CreatedBy = createdBy
        };

        _context.TipoPagos.Add(tipoPago);
        await _context.SaveChangesAsync();

        return MapearTipoPago(tipoPago);
    }

    public async Task<bool> EditarAsync(int id, TipoPagoEditarDto dto, int? updatedBy = null)
    {
        var tipoPago = await _context.TipoPagos
            .FirstOrDefaultAsync(t => t.TpaId == id && t.DeletedAt == null);

        if (tipoPago == null)
            return false;

        var nombre = dto.TpaNombre.Trim();

        var existe = await _context.TipoPagos
            .AnyAsync(t => t.TpaId != id && t.TpaNombre == nombre && t.DeletedAt == null);

        if (existe)
            throw new Exception("Ya existe otro tipo de pago con ese nombre.");

        tipoPago.TpaNombre = nombre;
        tipoPago.UpdatedAt = DateTime.Now;
        tipoPago.UpdatedBy = updatedBy;

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> EliminarAsync(int id, int? deletedBy = null)
    {
        var tipoPago = await _context.TipoPagos
            .FirstOrDefaultAsync(t => t.TpaId == id && t.DeletedAt == null);

        if (tipoPago == null)
            return false;

        var tieneFacturas = await _context.Facturas
            .AnyAsync(f => f.TpaId == id && f.DeletedAt == null);

        if (tieneFacturas)
            throw new Exception("No se puede eliminar el tipo de pago porque tiene facturas asociadas.");

        tipoPago.DeletedAt = DateTime.Now;
        tipoPago.DeletedBy = deletedBy;
        tipoPago.UpdatedAt = DateTime.Now;
        tipoPago.UpdatedBy = deletedBy;

        await _context.SaveChangesAsync();
        return true;
    }

    private static TipoPagoListadoDto MapearTipoPago(TipoPago tipoPago)
    {
        return new TipoPagoListadoDto
        {
            TpaId = tipoPago.TpaId,
            TpaNombre = tipoPago.TpaNombre,
            CreatedAt = tipoPago.CreatedAt,
            UpdatedAt = tipoPago.UpdatedAt
        };
    }
}